taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py

```diff
@@ -182,26 +231,44 @@ def _enqueue_dask_calls(self, dag, tag, rs, retries, enqueued_calls):
                 lazy = self.checkpoint_fn(dag, k)
 
             cat = self.category_name(sexpr)
-            if cat not in self._categories_known:
-                if self.resources:
-                    self.set_category_resources_max(cat, self.resources)
-                if self.resources_mode:
-                    self.set_category_mode(cat, self.resources_mode)
-
-                    if not self._categories_known:
-                        self.enable_monitoring()
-                self._categories_known.add(cat)
-
-            t = PythonTaskDask(self,
-                               dag, k, sexpr,
-                               category=cat,
-                               environment=self.environment,
-                               extra_files=self.extra_files,
-                               env_vars=self.env_vars,
-                               retries=retries,
-                               lazy_transfers=lazy)
-            t.set_tag(tag)  # tag that identifies this dag
-            enqueued_calls.append(t)
+            if self.task_mode == 'tasks':
+                if cat not in self._categories_known:
+                    if self.resources:
+                        self.set_category_resources_max(cat, self.resources)
+                    if self.resources_mode:
+                        self.set_category_mode(cat, self.resources_mode)
+
+                        if not self._categories_known:
+                            self.enable_monitoring()
+                    self._categories_known.add(cat)
+
+                t = PythonTaskDask(self,
+                                   dag, k, sexpr,
+                                   category=cat,
+                                   environment=self.environment,
+                                   extra_files=self.extra_files,
+                                   env_vars=self.env_vars,
+                                   retries=retries,
+                                   lazy_transfers=lazy)
+
+                if self.env_per_task:
+                    t.set_command('{} -e {} {}'.format('poncho_package_run', self.environment_name, t._command))
```

Comment:
Why not use `add_environment`?

Reply:
This is a means to have each task unpack its environment and activate it. And is intended for testing runs. I was unsure if there was a better way to do so asI saw some caching with `declare_poncho` and `add_environment` even when `cache=False`

Reply:
The problem is that not all environments are poncho. Some options:
- Add a flag to any of the `declare_*` that relate to environments that indicate whether they should be shared.
- Or, only share if the environment file was declared as cachable.

Reply:
See https://github.com/cooperative-computing-lab/cctools/issues/3643

Reply:
Looking at this code, I think that if you have two tasks running at the worker using the same poncho tarball, then they still share the environment, right? The only way they don't share it is if only one task runs at any time. If the worker gets an uninterrupted overlapping series of tasks, then I think that the same environment is used over and over.

Reply:
If that's true, and this code is only added for testing, then it might be much easier if the same tarball gets hard or symlinked with different names before declaring the tasks in your tests.  Otherwise we would need to modify how things are named according to the cache rules for little pay off.

Another option is to add an option to the run_in_env that poncho_package_create generates so that it re-expand the tarball every time, maybe with an environment variable or something like that.

Reply:
While testing I haven't had them share environments using this method (from looking at the plots). Each task would unpack and and activate the environment in its own sandbox. I think. I'll look into another method.